pwndbg/commands/hijack_fd.py

```diff
@@ -0,0 +1,283 @@
+                dup_asm,
+                shellcraft.syscall("SYS_close", regs["newfd"]),
+            ]
+        )
+    )
+
+
+@contextlib.asynccontextmanager
+async def exec_shellcode_with_stack(ec: pwndbg.dbg_mod.ExecutionController, blob, stack_size: int):
+    stack_start_diff = pwndbg.aglib.regs.sp
+    stack_start = stack_start_diff - stack_size
+    original_stack = pwndbg.aglib.memory.read(stack_start, stack_size)
+
+    try:
+        async with pwndbg.aglib.shellcode.exec_shellcode(
+            ec, blob, restore_context=True, disable_breakpoints=True
+        ):
+            stack_diff_size = stack_start_diff - pwndbg.aglib.regs.sp
+
+            # Make sure stack is not corrupted somehow
+            assert not (
+                stack_diff_size > stack_size
+            ), f"stack is probably corrupted size_current=f{stack_diff_size} size_max_want={stack_size}"
+
+            yield
+    finally:
+        pwndbg.aglib.memory.write(stack_start, original_stack)
+
+
+parser = argparse.ArgumentParser(
+    formatter_class=argparse.RawTextHelpFormatter,
+    description="""Replace file descriptors of a debugged process.
+
+The new file descriptor can point to:
+- a file
+- a pipe
+- a socket
+- a device, etc.
+
+Examples:
+1. Redirect STDOUT to a file:
+   `hijack-fd 1 /dev/null`
+
+2. Redirect STDERR to a socket:
+   `hijack-fd 2 tcp://localhost:8888`
+""",
+)
+
+parser.add_argument(
+    "fdnum",
+    help="File descriptor (FD) number to be replaced with the specified new socket or file.",
+    type=int,
+)
+
+
+class ParsedSocket(NamedTuple):
+    protocol: Literal["tcp", "udp"]
+    ip_version: Literal["ipv4", "ipv6"]
+    address: str
+    port: int
+
+
+def parse_socket(url: str) -> ParsedSocket:
+    if "://" in url:
+        parsed = urlparse(url)
+    else:
+        parsed = ParseResult("", url, "", "", "", "")
+
+    scheme_info = parsed.scheme.split("+", 1)
+
+    selected_protocol: Literal["tcp", "udp"] = "tcp"
+    selected_ip_protocol: Literal["ipv4", "ipv6"] | None = None
+    if parsed.scheme:
+        for any_value in scheme_info:
+            if any_value in ("tcp", "udp"):
+                selected_protocol = any_value
+            elif any_value in ("ipv4", "ipv6"):
+                selected_ip_protocol = any_value
+
+    domain_or_ip = parsed.hostname
+    if not domain_or_ip:
+        raise argparse.ArgumentTypeError("Domain or IP is required")
+
+    port = parsed.port
+    if not port:
+        raise argparse.ArgumentTypeError("Port is required")
+
+    protocol_ordered = [
+        ("ipv4", socket.AF_INET),
+        ("ipv6", socket.AF_INET6),
+    ]
+
+    found_ip_protocol: Literal["ipv4", "ipv6"] | None = None
+    address_ipv4_or_ipv6: str = ""
+    for family_name, family_const in protocol_ordered:
+        if selected_ip_protocol and selected_ip_protocol != family_name:
+            continue
+
+        try:
+            ips = socket.getaddrinfo(domain_or_ip, None, family_const)
```

Comment:
might be nice to add a comment what we do here (does it make sense to use addrinfo if we have an ip?)

Reply:
yeah, it resolves domain to ip, and if you put just `ip` then you know ip version ;) You don't need implement separate method like: `if is_ipv4 / ipv_6` statements